tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+
+			// Create the "live" event we are going to insert our backfilled events next to
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 2)
+			eventBefore := eventsBefore[len(eventsBefore)-1]
+			timeAfterEventBefore := time.Now()
+
+			numHistoricalMessages := 6
+			// wait X number of ms to ensure that the timestamp changes enough for each of the messages we try to backfill later
+			time.Sleep(time.Duration(numHistoricalMessages) * TimeBetweenMessages)
+
+			// Create some events after.
+			// Fill up the buffer so we have to scrollback to the inserted history later
+			eventsAfter := createMessagesInRoom(t, alice, roomID, 2)
+
+			// Register and join the virtual user
+			ensureVirtualUserRegistered(t, as, virtualUserLocalpart)
+
+			// Insert the most recent chunk of backfilled history
+			backfillRes := backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore.Add(TimeBetweenMessages*3),
+				"",
+				3,
+				// Status
+				200,
+			)
+			_, historicalEvents := getEventsFromBatchSendResponse(t, backfillRes)
+			nextChunkID := getNextChunkIdFromBatchSendResponse(t, backfillRes)
+
+			// Insert another older chunk of backfilled history from the same user.
+			// Make sure the meta data and joins still work on the subsequent chunk
+			backfillRes2 := backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				nextChunkID,
+				3,
+				// Status
+				200,
+			)
+			_, historicalEvents2 := getEventsFromBatchSendResponse(t, backfillRes2)
+
+			var expectedMessageOrder []string
+			expectedMessageOrder = append(expectedMessageOrder, eventsBefore...)
+			expectedMessageOrder = append(expectedMessageOrder, historicalEvents2...)
+			expectedMessageOrder = append(expectedMessageOrder, historicalEvents...)
+			expectedMessageOrder = append(expectedMessageOrder, eventsAfter...)
+			// Order events from newest to oldest
+			expectedMessageOrder = reversed(expectedMessageOrder)
+
+			messagesRes := alice.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+				"dir":   []string{"b"},
+				"limit": []string{"100"},
+			}))
+			messsageResBody := client.ParseJSON(t, messagesRes)
+			eventIDsFromResponse := getEventIDsFromResponseBody(t, messsageResBody)
+			// Since the original body can only be read once, create a new one from the body bytes we just read
+			messagesRes.Body = ioutil.NopCloser(bytes.NewBuffer(messsageResBody))
+
+			// Copy the array by value so we can modify it as we iterate in the foreach loop.
+			// We save the full untouched `expectedMessageOrder` for use in the log messages
+			workingExpectedMessageOrder := expectedMessageOrder
+
+			must.MatchResponse(t, messagesRes, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONArrayEach("chunk", func(r gjson.Result) error {
```

Comment:
This isn't a good assertion because if `chunk` is empty this will pass. Do length checks on `eventIDsFromResponse` prior to calling this.